### PR TITLE
Improve `PerMessageDeflateClientExtensionHandler`

### PR DESCRIPTION
**Pull request by ksmr**
Motivation:

The `PerMessageDeflateClientExtensionHandler` has the following strange behaviors currently:

* The `requestedServerNoContext` parameter doesn't actually add the `server_no_context_takeover` parameter to the client offer; instead it depends on the requested server window size.
* The handshake will fail if the server responds with a `server_no_context_takeover` parameter and `requestedServerNoContext` is false. According to RFC 7692 (7.1.1.1) the server may do this, and this means that to cover both cases one needs to use two handshakers in the channel pipeline: one with `requestedServerNoContext = true` and one with `requestedServerNoContext = false`.
* The value of the `server_max_window_bits` parameter in the server response is never checked (should be between 8 and 15). And the value of `client_max_window_bits` is checked only in the branch handling the server window parameter.

Modification:

* Add the `server_no_context_takeover` parameter if `requestedServerNoContext` is true.
* Accept a server handshake response which includes the server no context takeover parameter even if we did not request it.
* Check the values of the client and server window size in their respective branches and fail the handshake if they are out of bounds.

Result:

There will be no need to use two handshakers in the pipeline to be lenient in what handshakes are accepted.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ksmr can you please also add a unit test ?

**Comment by ksmr** [comment]
I've added unit tests that are fixed by the changes. :+1:

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateClientExtensionHandshakerTest.java [hunk 1]
```diff
@@ -130,6 +130,63 @@ public void testCustomHandshake() {
...
+        extension = handshaker.handshakeExtension(new WebSocketExtensionData(PERMESSAGE_DEFLATE_EXTENSION, parameters));
+
+        // Test that handshake fails when parameters are invalid
+        assertNull(extension);
+    }
+
+    @Test
+    public void testServerNoContextTakeover() {
+        WebSocketClientExtension extension;
+        Map<String, String> parameters;
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateClientExtensionHandshakerTest.java:165]: remove these two and just define them when you actually create the objects

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Review by slandelle — APPROVED** [review]
Great, thanks!

**Comment by normanmaurer** [comment]
@ksmr thanks a lot!

**Pull request merged**